File: RegexTester/ProjectOptions.vb
Imports System
Imports System.IO
Imports System.Text.RegularExpressions
Imports System.Xml.Serialization

Namespace RegexTester
    Public Class ProjectOptions
        ' Methods
        Public Sub ClearChanges()
            Me.LoadValues = DirectCast(Me.MemberwiseClone, ProjectOptions)
        End Sub

        Public Shared Function Load(ByVal fileName As String) As ProjectOptions
            Using stream As FileStream = New FileStream(fileName, FileMode.Open)
                Dim serializer As New XmlSerializer(GetType(ProjectOptions))
                Dim options2 As ProjectOptions = DirectCast(serializer.Deserialize(stream), ProjectOptions)
                options2.RegexFile = fileName
                options2.ClearChanges
                Return options2
            End Using
        End Function

        Public Sub Save(ByVal fileName As String)
            Using stream As FileStream = New FileStream(fileName, FileMode.Create)
                New XmlSerializer(GetType(ProjectOptions)).Serialize(DirectCast(stream, Stream), Me)
                Me.RegexFile = fileName
                Me.ClearChanges
            End Using
        End Sub


        ' Properties
        Public ReadOnly Property HasChanged As Boolean
            Get
                Return (((((Me.RegexText <> Me.LoadValues.RegexText) OrElse (Me.ReplaceText <> Me.LoadValues.ReplaceText)) OrElse ((Me.SourceText <> Me.LoadValues.SourceText) OrElse (Me.Command <> Me.LoadValues.Command))) OrElse (((Me.RegexOptions <> Me.LoadValues.RegexOptions) OrElse (Me.Detail <> Me.LoadValues.Detail)) OrElse ((Me.MaxMatches <> Me.LoadValues.MaxMatches) OrElse (Me.Sort <> Me.LoadValues.Sort)))) OrElse ((Me.Format <> Me.LoadValues.Format) OrElse (Me.WordWrap <> Me.LoadValues.WordWrap)))
            End Get
        End Property


        ' Fields
        <XmlIgnore> _
        Public RegexFile As String
        Public RegexName As String = ""
        Public RegexDescription As String = ""
        Public RegexText As String = ""
        Public ReplaceText As String = ""
        Public SourceText As String = ""
        Public Command As Command = Command.Find
        Public RegexOptions As RegexOptions
        Public Detail As DetailOption = DetailOption.Groups
        Public MaxMatches As Integer = &H3E8
        Public Sort As SortOption = SortOption.Position
        Public WordWrap As Boolean = True
        Public Format As FormatOption = FormatOption.Auto
        Public IncludeEmptyGroups As Boolean = True
        Public Language As LanguageOption = LanguageOption.VisualBasic
        Public VerbatimStrings As Boolean = False
        Public InstanceMethods As Boolean = True
        Public AssumeImports As Boolean = True
        Public GenerateLoop As Boolean = True
        Public IncludeComment As Boolean = True
        Public CopyCodeOnExit As Boolean = True
        <XmlIgnore> _
        Private LoadValues As ProjectOptions
    End Class
End Namespace

File: RegexTester/ProjectOptions.cs
namespace RegexTester
{
    using System;
    using System.IO;
    using System.Text.RegularExpressions;
    using System.Xml.Serialization;

    public class ProjectOptions
    {
        [XmlIgnore]
        public string RegexFile;
        public string RegexName = "";
        public string RegexDescription = "";
        public string RegexText = "";
        public string ReplaceText = "";
        public string SourceText = "";
        public RegexTester.Command Command = RegexTester.Command.Find;
        public System.Text.RegularExpressions.RegexOptions RegexOptions;
        public DetailOption Detail = DetailOption.Groups;
        public int MaxMatches = 0x3e8;
        public SortOption Sort = SortOption.Position;
        public bool WordWrap = true;
        public FormatOption Format = FormatOption.Auto;
        public bool IncludeEmptyGroups = true;
        public LanguageOption Language = LanguageOption.VisualBasic;
        public bool VerbatimStrings = false;
        public bool InstanceMethods = true;
        public bool AssumeImports = true;
        public bool GenerateLoop = true;
        public bool IncludeComment = true;
        public bool CopyCodeOnExit = true;
        [XmlIgnore]
        private ProjectOptions LoadValues;

        public void ClearChanges()
        {
            this.LoadValues = (ProjectOptions) this.MemberwiseClone();
        }

        public static ProjectOptions Load(string fileName)
        {
            using (FileStream stream = new FileStream(fileName, FileMode.Open))
            {
                XmlSerializer serializer = new XmlSerializer(typeof(ProjectOptions));
                ProjectOptions options2 = (ProjectOptions) serializer.Deserialize(stream);
                options2.RegexFile = fileName;
                options2.ClearChanges();
                return options2;
            }
        }

        public void Save(string fileName)
        {
            using (FileStream stream = new FileStream(fileName, FileMode.Create))
            {
                new XmlSerializer(typeof(ProjectOptions)).Serialize((Stream) stream, this);
                this.RegexFile = fileName;
                this.ClearChanges();
            }
        }

        public bool HasChanged
        {
            get
            {
                return (((((this.RegexText != this.LoadValues.RegexText) || (this.ReplaceText != this.LoadValues.ReplaceText)) || ((this.SourceText != this.LoadValues.SourceText) || (this.Command != this.LoadValues.Command))) || (((this.RegexOptions != this.LoadValues.RegexOptions) || (this.Detail != this.LoadValues.Detail)) || ((this.MaxMatches != this.LoadValues.MaxMatches) || (this.Sort != this.LoadValues.Sort)))) || ((this.Format != this.LoadValues.Format) || (this.WordWrap != this.LoadValues.WordWrap)));
            }
        }
    }
}

